### PR TITLE
chore: more factoring

### PR DESCRIPTION
**Pull request by peternhale**

### CODE REVIEW

File: test/utilities/extensionUtils.ts [hunk 1]
```diff
@@ -24,7 +25,8 @@ export type ExtensionId =
   | 'salesforcedx-vscode-apex-replay-debugger'
   | 'salesforcedx-vscode-lightning'
   | 'salesforcedx-vscode-lwc'
-  | 'salesforcedx-vscode-visualforce';
+  | 'salesforcedx-vscode-visualforce'
+  | 'salesforce.salesforce-vscode-slds';
```
Comment (CristiCanizales) [inline test/utilities/extensionUtils.ts:29]: Why do we need to have this extension here? And at the same time removing the validation of the commands in anInitialSuite.e2e.ts?

File: test/specs/debugLwcTests.e2e.ts [hunk 2]
```diff
@@ -14,7 +14,7 @@ import util from 'util';
 import { fail } from 'assert';
 
 import { Key } from 'webdriverio';
-const CMD_KEY = process.platform === 'darwin' ? Key.Command : Key.Control;
+const CMD_KEY = process.platform === 'darwin' ? Key.Command : Key.Control;import { Duration } from '@salesforce/kit';
```
Comment (CristiCanizales) [inline test/specs/debugLwcTests.e2e.ts:17]:
```suggestion
const CMD_KEY = process.platform === 'darwin' ? Key.Command : Key.Control;
import { Duration } from '@salesforce/kit';
```

File: test/specs/orgBrowser.e2e.ts [hunk 3]
```diff
@@ -156,7 +169,7 @@ describe('Org Browser', async () => {
     utilities.log(`${testSetup.testSuiteSuffixName} - Retrieve and Open Source`);
     const myClassLabelEl = await utilities.findElementByText('div', 'aria-label', 'MyClass');
     myClassLabelEl.click();
-    utilities.pause(2);
+    utilities.pause(Duration.seconds(2));
```
Comment (CristiCanizales) [inline test/specs/orgBrowser.e2e.ts:172]:
```suggestion
    await utilities.pause(Duration.seconds(2));
```

**Review by CristiCanizales — CHANGES_REQUESTED** [review]
Looks great! Just some questions:

1. what is the benefit of using Duration.seconds(x) instead of x?
2. why do we include seconds in clearOutputView() in some places and not in others?

File: test/specs/manifestBuilder.e2e.ts [hunk 4]
```diff
@@ -120,11 +121,11 @@ describe('Manifest Builder', async () => {
     const workbench = await browser.getWorkbench();
     await utilities.getTextEditor(workbench, 'manifest.xml');
     // Clear output before running the command
-    await utilities.runCommandFromCommandPrompt(workbench, 'View: Clear Output', 1);
+    await utilities.clearOutputView(Duration.seconds(2));
```
Comment (daphne-sfdc) [inline test/specs/manifestBuilder.e2e.ts:124]: It looks like `clearOutputView()` sometimes has a duration of 1 second and sometimes has a duration of 2 seconds.  Would it make sense to standardize it to 1 second throughout the code, especially since that's the default timeout?

File: test/specs/manifestBuilder.e2e.ts [hunk 5]
```diff
@@ -120,11 +121,11 @@ describe('Manifest Builder', async () => {
...
     await utilities.runCommandFromCommandPrompt(
       workbench,
       'SFDX: Retrieve Source in Manifest from Org',
-      1
+      Duration.seconds(2)
```
Comment (daphne-sfdc) [inline test/specs/manifestBuilder.e2e.ts:128]: Why was this changed from 1 second to 2 seconds?  Is there a flapper that happens when it's 1 second?

File: test/utilities/extensionUtils.ts (hunk 1 above)
Comment (daphne-sfdc) [reply test/utilities/extensionUtils.ts:29]: I removed the validation of the commands in anInitialSuite.e2e.ts because SLDS is installed as part of the Salesforce Extension Pack and we no longer install the packs in E2E.  In this case I agree with Cristi that Line 29 can be removed.

File: test/utilities/miscellaneous.ts [hunk 6]
```diff
@@ -8,16 +8,18 @@
...
 import { executeQuickPick } from './commandPrompt.ts';
 import { notificationIsPresentWithTimeout } from './notifications.ts';
+import { Duration } from '@salesforce/kit';
 import path from 'path';
+import { PredicateWithTimeout } from './predicates.ts';
 
-export const FIVE_MINUTES = 5 * 60;
-export const TEN_MINUTES = 10 * 60;
+export const FIVE_MINUTES = Duration.minutes(5);
+export const TEN_MINUTES = Duration.minutes(10);
```
Comment (daphne-sfdc) [inline test/utilities/miscellaneous.ts:19]: 😀

File: test/utilities/outputView.ts [hunk 7]
```diff
@@ -58,68 +52,59 @@ export async function attemptToFindOutputPanelText(
...
-    if (outputViewText.includes(searchString)) {
-      return outputViewText;
+  browser.waitUntil(
+    async () => {
+      const outputViewText = await getOutputViewText();
+      return outputViewText.includes(searchString);
+    },
+    {
+      timeout: 1000 * attempts, // Convert attempts to milliseconds assuming each attempt is 1 second
+      interval: 500, // Check every second
```
Comment (daphne-sfdc) [inline test/utilities/outputView.ts:62]: Should the comment be "Check every half second" instead?

Comment (daphne-sfdc) [inline test/utilities/outputView.ts:62]: Why is the interval 500 instead of 1000?

File: test/utilities/workbench.ts [hunk 8]
```diff
@@ -43,24 +44,25 @@ export async function showExplorerView(): Promise<void> {
...
   await executeQuickPick('View: Reset Zoom', wait);
 }
 
-async function handlePredicateOrWait(predicateOrWait: PredicateWithTimeout | number, prompt: unknown) {
-  if (typeof predicateOrWait === 'number') {
-    if (predicateOrWait > 0) {
+async function handlePredicateOrWait(predicateOrWait: PredicateWithTimeout | Duration, prompt: unknown) {
+  log('handlePredicateOrWait')
+  if (isDuration(predicateOrWait)) {
+    if (predicateOrWait && predicateOrWait.milliseconds > 0) {
```
Comment (daphne-sfdc) [inline test/utilities/workbench.ts:62]:
```suggestion
    if (predicateOrWait?.milliseconds > 0) {
```

File: test/utilities/extensionUtils.ts (hunk 1 above)
Comment (peternhale) [reply test/utilities/extensionUtils.ts:29]: I'll remove the unneeded extension.

File: test/specs/manifestBuilder.e2e.ts (hunk 4 above)
Comment (peternhale) [reply test/specs/manifestBuilder.e2e.ts:124]: I'll change it to 1 second

File: test/specs/manifestBuilder.e2e.ts (hunk 5 above)
Comment (peternhale) [reply test/specs/manifestBuilder.e2e.ts:128]: here too

File: test/utilities/outputView.ts (hunk 7 above)
Comment (peternhale) [reply test/utilities/outputView.ts:62]: reduce overall wait time

Comment (peternhale) [reply test/utilities/outputView.ts:62]: I'll fix the comment

File: test/utilities/workbench.ts (hunk 8 above)
Comment (peternhale) [reply test/utilities/workbench.ts:62]: The type assertion on line 62 guarantees predicateOrWait is not undefined, but is also of type Duration, so the safe navigation is not needed either.

**Comment by peternhale** [comment]
> Looks great! Just some questions:
> 
> 1. what is the benefit of using Duration.seconds(x) instead of x?
> 2. why do we include seconds in clearOutputView() in some places and not in others?

1. We all know that the calls to sleep in pause require milliseconds, yet pause function's parameter is just a number that everyone should know that the unit is seconds. Changing the type to Duration takes away the need to know how the parameter is used within pause.  It also gives the test writer the ability to express the value in anyway they would like. Granted the most common use is using seconds, but now we can express the values in any time unit we would like. Duration.milliseconds(1000), Duration.seconds(1),  Duration.minutes(10). And when consumed, we can use the unit suitable for the need, `sleep(wait.milliseconds)`. Also, the change now allows us to use a unit smaller than a second, `pause(Duration.milliseconds(500))` without having to deal with unit conversions in the consuming function.
2. I believe I set a default duration of 1 second, so not providing uses the default.

File: test/specs/pushAndPull.e2e.ts [hunk 9]
```diff
@@ -94,70 +97,67 @@ describe('Push and Pull', async () => {
...
+    await utilities.clearOutputView(Duration.seconds(2));
 
     // Now push
-    await utilities.runCommandFromCommandPrompt(workbench, 'SFDX: Push Source to Default Org', 5);
-
-    await verifyPushSuccess(workbench);
+    await utilities.executeQuickPick('SFDX: Push Source to Default Org', Duration.seconds(5));
 
     // Check the output.
-    await verifyPushAndPullOutputText(workbench, 'Push', 'to');
```
Comment (daphne-sfdc) [inline test/specs/pushAndPull.e2e.ts:127]: Why don't we want to verify the output text after the push anymore?

File: test/specs/pushAndPull.e2e.ts [hunk 10]
```diff
@@ -94,70 +97,67 @@ describe('Push and Pull', async () => {
...
     const textEditor = await utilities.getTextEditor(workbench, 'ExampleApexClass1.cls');
     await textEditor.setTextAtLine(3, '        // sample comment');
 
     // Push the file.
-    await utilities.runCommandFromCommandPrompt(workbench, 'SFDX: Push Source to Default Org', 5);
+    await utilities.executeQuickPick('SFDX: Push Source to Default Org', Duration.seconds(5));
 
     await verifyPushSuccess(workbench);
 
-    await verifyPushAndPullOutputText(workbench, 'Push', 'to');
```
Comment (daphne-sfdc) [inline test/specs/pushAndPull.e2e.ts:145]: Same here - why is the validation removed?

File: test/specs/pushAndPull.e2e.ts [hunk 11]
```diff
@@ -183,87 +183,78 @@ describe('Push and Pull', async () => {
...
+    await utilities.clearOutputView(Duration.seconds(2));
 
     // And pull again.
-    await utilities.runCommandFromCommandPrompt(workbench, 'SFDX: Pull Source from Default Org', 5);
-    await verifyPullSuccess(workbench);
-
+    await utilities.executeQuickPick('SFDX: Pull Source from Default Org', Duration.seconds(5));
     // Check the output.
-    outputPanelText = await verifyPushAndPullOutputText(workbench, 'Pull', 'from');
-    expect(outputPanelText).not.toContain('Created  Admin');
```
Comment (daphne-sfdc) [inline test/specs/pushAndPull.e2e.ts:211]: And here - validation got removed.

File: test/specs/pushAndPull.e2e.ts [hunk 12]
```diff
@@ -183,87 +183,78 @@ describe('Push and Pull', async () => {
...
     await textEditor.setTextAtLine(3, '        // sample comment for the pull test');
     // Don't save the file just yet.
 
     // Pull the file.
-    await utilities.runCommandFromCommandPrompt(workbench, 'SFDX: Pull Source from Default Org', 5);
-    await verifyPullSuccess(workbench);
-
+    await utilities.executeQuickPick('SFDX: Pull Source from Default Org', Duration.seconds(5));
     // Check the output.
-    await verifyPushAndPullOutputText(workbench, 'Pull', 'from');
```
Comment (daphne-sfdc) [inline test/specs/pushAndPull.e2e.ts:230]: Also here - the validation is gone.

File: test/specs/pushAndPull.e2e.ts [hunk 13]
```diff
@@ -183,87 +183,78 @@ describe('Push and Pull', async () => {
...
 
     // Now save the file.
     const textEditor = await utilities.getTextEditor(workbench, 'ExampleApexClass1.cls');
     await textEditor.save();
 
     // An now pull the changes.
-    await utilities.runCommandFromCommandPrompt(workbench, 'SFDX: Pull Source from Default Org', 5);
+    await utilities.executeQuickPick('SFDX: Pull Source from Default Org', Duration.seconds(5));
     await verifyPullSuccess(workbench);
-    await verifyPushAndPullOutputText(workbench, 'Pull', 'from');
```
Comment (daphne-sfdc) [inline test/specs/pushAndPull.e2e.ts:246]: Here too for the missing validation.

File: test/specs/runEinsteinForDeveloper.e2e.ts
Comment (daphne-sfdc) [inline test/specs/runEinsteinForDeveloper.e2e.ts:1]: Why is this file in our automation tests repo?  Doesn't E4D have their own repo with their E2E tests already?  (Or are we using it to be able to run their tests from our workflows?)

File: test/utilities/miscellaneous.ts [hunk 14]
```diff
@@ -81,14 +91,14 @@ export async function createCommand(
   folder: string,
   extension: string
 ): Promise<string | undefined> {
-  const workbench = await (await browser.getWorkbench()).wait();
-  await executeQuickPick('View: Clear Output', 1);
-  const inputBox = await executeQuickPick(`SFDX: Create ${type}`, 1);
+  const workbench = await getWorkbench();
+  await executeQuickPick('View: Clear Output', Duration.seconds(1));
```
Comment (daphne-sfdc) [inline test/utilities/miscellaneous.ts:95]: Import and use the helper function from `outputView.ts`

```suggestion
  await clearOutputView();
```

File: test/environmentSettings.ts [hunk 15]
```diff
@@ -35,7 +35,8 @@ export class EnvironmentSettings {
     // './specs/**/templates.e2e.ts',
     // './specs/**/trailApexReplayDebugger.e2e.ts',
     // './specs/**/visualforceLsp.e2e.ts',
-    // './specs/**/runEinsteinForDeveloper.e2e.ts'
+    // './specs/**/runEinsteinForDeveloper.e2e.ts',
```
Comment (daphne-sfdc) [inline test/environmentSettings.ts:38]: If we want to remove the `runEinsteinForDeveloper.e2e.ts` file from this repo, this line should also be removed.

File: test/specs/runEinsteinForDeveloper.e2e.ts
Comment (peternhale) [reply test/specs/runEinsteinForDeveloper.e2e.ts:1]: I don't know but I will remove it.

File: test/utilities/extensionUtils.ts [hunk 16]
```diff
@@ -12,6 +12,7 @@ import FastGlob from 'fast-glob';
 import { EnvironmentSettings } from '../environmentSettings.ts';
 import { exec } from 'child_process';
 import * as utilities from './index.ts';
+import { Duration } from '@salesforce/kit';
 
 export type ExtensionId =
```
[the remark this answers is not in the document]
Comment (peternhale) [reply test/utilities/extensionUtils.ts:17]: I recall discussions around writing simple, will our extension packs start, so maybe it should stay here until such time where we know the scope of the "will our extension packs start" tests.

File: test/specs/pushAndPull.e2e.ts (hunk 10 above)
Comment (daphne-sfdc) [reply test/specs/pushAndPull.e2e.ts:145]: Is this actually meant to be removed?  I see that this one is the only `verifyPushAndPullOutputText()` that is missing now.

**Review by daphne-sfdc — APPROVED** [review]
Approved! 🎉 

Thanks Pete for all the great refactoring work done in this PR! 😀 

E2E test run here: https://github.com/forcedotcom/salesforcedx-vscode/actions/runs/9944040002 ✅
